File: sdk/swift/Sources/TokenPayID/TpidConfig.swift
import Foundation

/// Configuration for TOKEN PAY ID native widget.
public struct TpidConfig: Sendable, Equatable {
    public let clientId: String
    public let redirectURI: URL
    public let scopes: [String]
    public let issuer: URL
    public let theme: Theme
    public let language: Language
    public let presentation: Presentation
    public let prefillEmail: String?
    public let allowRegister: Bool
    public let allowPasskey: Bool
    public let allowRecovery: Bool
    public let pinTLSCertificates: Bool
    public let enableTelemetry: Bool
    public let brandingOverride: Branding?

    public init(
        clientId: String,
        redirectURI: URL,
        scopes: [String] = ["openid", "profile", "email"],
        issuer: URL = URL(string: "https://id.tokenpay.space")!,
        theme: Theme = .auto,
        language: Language = .system,
        presentation: Presentation = .sheet,
        prefillEmail: String? = nil,
        allowRegister: Bool = true,
        allowPasskey: Bool = true,
        allowRecovery: Bool = true,
        pinTLSCertificates: Bool = true,
        enableTelemetry: Bool = true,
        brandingOverride: Branding? = nil
    ) {
        precondition(clientId.hasPrefix("tpid_pk_"), "clientId must start with 'tpid_pk_'")
        precondition(!scopes.isEmpty, "at least one scope required")
        precondition(issuer.scheme == "https", "issuer must be HTTPS")
        self.clientId = clientId
        self.redirectURI = redirectURI
        self.scopes = scopes
        self.issuer = issuer
        self.theme = theme
        self.language = language
        self.presentation = presentation
        self.prefillEmail = prefillEmail
        self.allowRegister = allowRegister
        self.allowPasskey = allowPasskey
        self.allowRecovery = allowRecovery
        self.pinTLSCertificates = pinTLSCertificates
        self.enableTelemetry = enableTelemetry
        self.brandingOverride = brandingOverride
    }

    public enum Theme: String, Sendable, Codable { case light, dark, auto }

    public enum Language: String, Sendable, Codable {
        case system, ru, en, zh
        public static func fromCode(_ code: String?) -> Language {
            Language(rawValue: code ?? "system") ?? .system
        }
    }

    public enum Presentation: Sendable, Equatable {
        case sheet         // modal sheet on iOS, sheet on macOS
        case fullScreen    // fullScreenCover on iOS
        case embedded      // caller embeds in own view tree
    }

    public struct Branding: Sendable, Equatable, Codable {
        public let logoURL: URL?
        public let primaryColorHex: String?
        public let appName: String?
        public let appDomain: String?
        public init(logoURL: URL? = nil, primaryColorHex: String? = nil, appName: String? = nil, appDomain: String? = nil) {
            self.logoURL = logoURL
            self.primaryColorHex = primaryColorHex
            self.appName = appName
            self.appDomain = appDomain
        }
    }

    public var scopeString: String { scopes.joined(separator: " ") }
}

File: sdk/android/tokenpay-id-sdk/src/main/java/space/tokenpay/id/TpidConfig.kt
package space.tokenpay.id

/**
 * Configuration for TOKEN PAY ID native widget.
 *
 * @property clientId Your public client key (`tpid_pk_...`). Required.
 * @property redirectUri OAuth redirect URI. Must match the one registered for your client.
 *           Typical value: `"${applicationId}:/auth/callback"`.
 * @property scopes OAuth scopes requested. Default: `[openid, profile, email]`.
 * @property issuer Base URL of the authorization server. Default: `https://id.tokenpay.space`.
 * @property theme Widget color theme. Default: [Theme.AUTO] (follows system).
 * @property language UI language. Default: [Language.SYSTEM].
 * @property presentation How the widget is shown. Default: [Presentation.MODAL].
 * @property prefillEmail Pre-fill the email field (for multi-account flows).
 * @property allowRegister Whether the widget shows the "Create account" option. Default: true.
 * @property allowPasskey Enable passkey/biometric authentication. Default: true.
 * @property allowRecovery Enable recovery flow in the widget. Default: true.
 * @property allowScreenshots Disable FLAG_SECURE for screenshots (debug only). Default: false.
 * @property pinTlsCertificates Enable TLS certificate pinning. Default: true. Disable only for dev.
 * @property enableTelemetry Send anonymous funnel events to TOKEN PAY. Default: true.
 * @property brandingOverride Force custom branding (normally fetched from server based on clientId).
 */
data class TpidConfig(
    val clientId: String,
    val redirectUri: String,
    val scopes: List<String> = listOf("openid", "profile", "email"),
    val issuer: String = "https://id.tokenpay.space",
    val theme: Theme = Theme.AUTO,
    val language: Language = Language.SYSTEM,
    val presentation: Presentation = Presentation.MODAL,
    val prefillEmail: String? = null,
    val allowRegister: Boolean = true,
    val allowPasskey: Boolean = true,
    val allowRecovery: Boolean = true,
    val allowScreenshots: Boolean = false,
    val pinTlsCertificates: Boolean = true,
    val enableTelemetry: Boolean = true,
    val brandingOverride: Branding? = null,
) {
    init {
        require(clientId.startsWith("tpid_pk_")) { "clientId must start with 'tpid_pk_'" }
        require(redirectUri.isNotBlank()) { "redirectUri required" }
        require(scopes.isNotEmpty()) { "at least one scope required" }
        require(issuer.startsWith("https://")) { "issuer must be HTTPS" }
    }

    enum class Theme { LIGHT, DARK, AUTO }

    enum class Language(val code: String) {
        SYSTEM("system"),
        RUSSIAN("ru"),
        ENGLISH("en"),
        CHINESE("zh");
        companion object {
            fun fromCode(code: String?): Language = values().firstOrNull { it.code == code } ?: SYSTEM
        }
    }

    enum class Presentation { MODAL, FULLSCREEN, EMBEDDED }

    data class Branding(
        val logoUrl: String?,
        val primaryColor: String?,
        val appName: String?,
        val appDomain: String?,
    )

    internal val scopeString: String get() = scopes.joinToString(" ")
}
